### ISSUE TITLE
map is overwritten for non-StaticArrays.

### ISSUE THREAD

**Issue by andreasnoack** [JuliaArrays/StaticArrays.jl#282]
The problematic lines are https://github.com/JuliaArrays/StaticArrays.jl/blob/7276787a0b147135b758dd836b5951236e30558d/src/mapreduce.jl#L9-L11 The new behavior might be https://github.com/JuliaLang/julia/issues/21026. I'm not sure

This is on 
```julia
Julia Version 0.7.0-DEV.1327
Commit 104dce7fcf* (2017-08-11 16:01 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT NO_AFFINITY HASWELL)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
Environment:
```

**Comment by tkoolen** [comment]
Ref: https://github.com/JuliaLang/julia/issues/21026#issuecomment-322873794.

**Comment by andyferris** [comment]
Thanks. We'll need to dream up a workaround for this little "hack" to express such a set of types.

For anyone finding this (and issues like this): It's likely that at some point it will become too difficult to target both v0.6 and v0.7. Given Julia changes like https://github.com/JuliaLang/julia/pull/23168 I'm very tempted to just say "StaticArrays should not be considered as suitable for "real work" on v0.7 until much closer to to the release time". I feel `master` should target v0.6 (with "feature" improvements like LU decompositions) and we merge any v0.7 fixes which are compatible with v0.6. I'm hopeful we can find a simple fix for this issue.

**Comment by c42f** [comment]
Hmm, this is unfortunate, though it was always a bit of a hack relying on this peculiarity.

The really nasty thing IIRC was if you try to do this in another way (by directly writing out various function signatures), the number of ambiguities quickly explodes in your face.

Another option might be to refactor the `map` internals to look a bit more like `broadcast`, as `Broadcast.containertype()` was pretty nice to hook into. Obviously this would be changes to `Base`...

**Comment by martinholters** [comment]
Fixed by #296. (Although the new behavior---only intercept `map` if the first argument (after the function) is a `StaticArray`--- still leaves something to be desired, this particular issue is fixed.)

**Comment by andyferris** [comment]
OK, let's just close this since the OP seems to be resolved - there are always many things which can be improved further in *StaticArrays* :)
